File: python/features/steps/dependency_steps.py
"""Behave steps for dependency scenarios."""

from __future__ import annotations

from behave import given, then, when

from dataclasses import dataclass
from pathlib import Path

from features.steps.shared import (
    build_issue,
    load_project_directory,
    read_issue_file,
    run_cli,
    write_issue_file,
)
from taskulus.dependencies import DependencyError, add_dependency, list_ready_issues
from taskulus.models import DependencyLink


@given('issue "{identifier}" depends on "{target}" with type "{dependency_type}"')
def given_issue_depends_on(
    context: object, identifier: str, target: str, dependency_type: str
) -> None:
    project_dir = load_project_directory(context)
    issue = build_issue(identifier, "Title", "task", "open", None, [])
    dependency = DependencyLink(target=target, type=dependency_type)
    issue = issue.model_copy(update={"dependencies": [dependency]})
    write_issue_file(project_dir, issue)


@then('issue "{identifier}" should depend on "{target}" with type "{dependency_type}"')
def then_issue_should_depend_on(
    context: object, identifier: str, target: str, dependency_type: str
) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert any(
        link.target == target and link.dependency_type == dependency_type
        for link in issue.dependencies
    )


@then(
    'issue "{identifier}" should not depend on "{target}" with type "{dependency_type}"'
)
def then_issue_should_not_depend_on(
    context: object, identifier: str, target: str, dependency_type: str
) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert not any(
        link.target == target and link.dependency_type == dependency_type
        for link in issue.dependencies
    )


@when('I run "tsk dep add tsk-child --blocked-by tsk-parent"')
def when_run_dep_add_blocked(context: object) -> None:
    from features.steps.shared import run_cli

    run_cli(context, "tsk dep add tsk-child --blocked-by tsk-parent")


@when('I run "tsk dep add tsk-left --relates-to tsk-right"')
def when_run_dep_add_relates(context: object) -> None:
    from features.steps.shared import run_cli

    run_cli(context, "tsk dep add tsk-left --relates-to tsk-right")


@when('I run "tsk dep add tsk-left --blocked-by tsk-right"')
def when_run_dep_add_blocked_left(context: object) -> None:
    run_cli(context, "tsk dep add tsk-left --blocked-by tsk-right")


@when('I run "tsk dep remove tsk-left --blocked-by tsk-right"')
def when_run_dep_remove(context: object) -> None:
    run_cli(context, "tsk dep remove tsk-left --blocked-by tsk-right")


@when('I run "tsk dep remove tsk-left --relates-to tsk-right"')
def when_run_dep_remove_relates(context: object) -> None:
    run_cli(context, "tsk dep remove tsk-left --relates-to tsk-right")


@when('I run "tsk dep add tsk-b --blocked-by tsk-a"')
def when_run_dep_add_cycle(context: object) -> None:
    run_cli(context, "tsk dep add tsk-b --blocked-by tsk-a")


@when('I run "tsk dep add tsk-a --blocked-by tsk-c"')
def when_run_dep_add_shared_downstream(context: object) -> None:
    run_cli(context, "tsk dep add tsk-a --blocked-by tsk-c")


@when('I run "tsk dep add tsk-missing --blocked-by tsk-parent"')
def when_run_dep_add_missing_issue(context: object) -> None:
    run_cli(context, "tsk dep add tsk-missing --blocked-by tsk-parent")


@when('I run "tsk ready"')
def when_run_ready(context: object) -> None:
    run_cli(context, "tsk ready")


@when('I run "tsk ready --local-only --no-local"')
def when_run_ready_conflict(context: object) -> None:
    run_cli(context, "tsk ready --local-only --no-local")


@when('I run "tsk ready --local-only"')
def when_run_ready_local_only(context: object) -> None:
    run_cli(context, "tsk ready --local-only")


@when('I run "tsk ready --no-local"')
def when_run_ready_no_local(context: object) -> None:
    run_cli(context, "tsk ready --no-local")


@when("ready issues are listed for a single project")
def when_ready_issues_listed_for_single_project(context: object) -> None:
    root = Path(context.working_directory)
    canonical_root = root.resolve()
    issues = list_ready_issues(
        root=canonical_root, include_local=True, local_only=False
    )
    context.ready_issue_ids = [issue.identifier for issue in issues]


@then('the ready list should contain "{identifier}"')
def then_ready_list_should_contain(context: object, identifier: str) -> None:
    ready_ids = getattr(context, "ready_issue_ids", [])
    assert identifier in ready_ids


@when('I run "tsk dep tree tsk-child"')
def when_run_dep_tree_child(context: object) -> None:
    run_cli(context, "tsk dep tree tsk-child")


@when('I run "tsk dep tree tsk-c --depth 1"')
def when_run_dep_tree_depth(context: object) -> None:
    run_cli(context, "tsk dep tree tsk-c --depth 1")


@when('I run "tsk dep tree tsk-root"')
def when_run_dep_tree_root(context: object) -> None:
    run_cli(context, "tsk dep tree tsk-root")


@when('I run "tsk dep tree tsk-missing"')
def when_run_dep_tree_missing(context: object) -> None:
    run_cli(context, "tsk dep tree tsk-missing")


@when('I run "tsk dep tree tsk-a"')
def when_run_dep_tree_a(context: object) -> None:
    run_cli(context, "tsk dep tree tsk-a")


@when('I run "tsk dep tree tsk-child --format json"')
def when_run_dep_tree_json(context: object) -> None:
    run_cli(context, "tsk dep tree tsk-child --format json")


@when('I run "tsk dep tree tsk-child --format dot"')
def when_run_dep_tree_dot(context: object) -> None:
    run_cli(context, "tsk dep tree tsk-child --format dot")


@when('I run "tsk dep tree tsk-child --format invalid"')
def when_run_dep_tree_invalid(context: object) -> None:
    run_cli(context, "tsk dep tree tsk-child --format invalid")


@given("a dependency tree with more than 25 nodes exists")
def given_large_dependency_tree(context: object) -> None:
    project_dir = load_project_directory(context)
    chain_length = 26
    for index in range(chain_length):
        identifier = "tsk-root" if index == 0 else f"tsk-node-{index}"
        issue = build_issue(identifier, f"Node {index}", "task", "open", None, [])
        if index < chain_length - 1:
            target = f"tsk-node-{index + 1}"
            issue = issue.model_copy(
                update={
                    "dependencies": [DependencyLink(target=target, type="blocked-by")]
                }
            )
        write_issue_file(project_dir, issue)


@when('I run "tsk dep add tsk-child"')
def when_run_dep_add_missing_target(context: object) -> None:
    run_cli(context, "tsk dep add tsk-child")


@when('I run "tsk dep remove tsk-child"')
def when_run_dep_remove_missing_target(context: object) -> None:
    run_cli(context, "tsk dep remove tsk-child")


@when('I run "tsk dep remove tsk-missing --blocked-by tsk-parent"')
def when_run_dep_remove_missing_issue(context: object) -> None:
    run_cli(context, "tsk dep remove tsk-missing --blocked-by tsk-parent")


@dataclass
class _DummyResult:
    exit_code: int
    stdout: str
    stderr: str


@when("I add an invalid dependency type")
def when_add_invalid_dependency_type(context: object) -> None:
    project_dir = load_project_directory(context)
    root = project_dir.parent
    try:
        add_dependency(root, "tsk-left", "tsk-right", "invalid-type")
    except DependencyError as error:
        context.result = _DummyResult(exit_code=1, stdout="", stderr=str(error))
        return
    context.result = _DummyResult(exit_code=0, stdout="", stderr="")


@then('issue "{identifier}" should have 1 dependency')
def then_issue_has_single_dependency(context: object, identifier: str) -> None:
    project_dir = load_project_directory(context)
    issue = read_issue_file(project_dir, identifier)
    assert len(issue.dependencies) == 1

File: rust/features/steps/dependency_steps.rs
use std::fs;
use std::path::PathBuf;

use chrono::{TimeZone, Utc};
use cucumber::{given, then, when};

use kanbus::dependencies::{add_dependency, list_ready_issues};
use kanbus::file_io::load_project_directory;
use kanbus::models::{DependencyLink, IssueData};

use crate::step_definitions::initialization_steps::KanbusWorld;

fn load_project_dir(world: &KanbusWorld) -> PathBuf {
    let cwd = world.working_directory.as_ref().expect("cwd");
    load_project_directory(cwd).expect("project dir")
}

fn write_issue_file(project_dir: &PathBuf, issue: &IssueData) {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{}.json", issue.identifier));
    let contents = serde_json::to_string_pretty(issue).expect("serialize issue");
    fs::write(issue_path, contents).expect("write issue");
}

fn read_issue_file(project_dir: &PathBuf, identifier: &str) -> IssueData {
    let issue_path = project_dir
        .join("issues")
        .join(format!("{identifier}.json"));
    let contents = fs::read_to_string(issue_path).expect("read issue");
    serde_json::from_str(&contents).expect("parse issue")
}

fn build_issue(identifier: &str) -> IssueData {
    let timestamp = Utc.with_ymd_and_hms(2026, 2, 11, 0, 0, 0).unwrap();
    IssueData {
        identifier: identifier.to_string(),
        title: "Title".to_string(),
        description: "".to_string(),
        issue_type: "task".to_string(),
        status: "open".to_string(),
        priority: 2,
        assignee: None,
        creator: None,
        parent: None,
        labels: Vec::new(),
        dependencies: Vec::new(),
        comments: Vec::new(),
        created_at: timestamp,
        updated_at: timestamp,
        closed_at: None,
        custom: std::collections::BTreeMap::new(),
    }
}

#[given(expr = "issue {string} depends on {string} with type {string}")]
fn given_issue_depends_on(
    world: &mut KanbusWorld,
    identifier: String,
    target: String,
    dependency_type: String,
) {
    let project_dir = load_project_dir(world);
    let mut issue = build_issue(&identifier);
    issue.dependencies.push(DependencyLink {
        target,
        dependency_type,
    });
    write_issue_file(&project_dir, &issue);
}

#[when("ready issues are listed for a single project")]
fn when_ready_issues_listed_single_project(world: &mut KanbusWorld) {
    let root = world.working_directory.as_ref().expect("cwd");
    let canonical = root.canonicalize().unwrap_or_else(|_| root.clone());
    let issues = list_ready_issues(&canonical, true, false).expect("ready list");
    world.ready_issue_ids = Some(issues.into_iter().map(|issue| issue.identifier).collect());
}

#[then(expr = "the ready list should contain {string}")]
fn then_ready_list_contains(world: &mut KanbusWorld, identifier: String) {
    let ids = world.ready_issue_ids.as_ref().expect("ready ids not set");
    assert!(ids.contains(&identifier));
}

#[when("I add an invalid dependency type")]
fn when_add_invalid_dependency(world: &mut KanbusWorld) {
    let root = world.working_directory.as_ref().expect("cwd");
    match add_dependency(root, "kanbus-left", "kanbus-right", "invalid-type") {
        Ok(_) => {
            world.exit_code = Some(0);
            world.stderr = Some(String::new());
        }
        Err(error) => {
            world.exit_code = Some(1);
            world.stderr = Some(error.to_string());
        }
    }
    world.stdout = Some(String::new());
}

#[given("a dependency tree with more than 25 nodes exists")]
fn given_large_dependency_tree(world: &mut KanbusWorld) {
    let project_dir = load_project_dir(world);
    let chain_length = 26;
    for index in 0..chain_length {
        let identifier = if index == 0 {
            "kanbus-root".to_string()
        } else {
            format!("kanbus-node-{}", index)
        };
        let mut issue = build_issue(&identifier);
        if index < chain_length - 1 {
            let target = format!("kanbus-node-{}", index + 1);
            issue.dependencies.push(DependencyLink {
                target,
                dependency_type: "blocked-by".to_string(),
            });
        }
        write_issue_file(&project_dir, &issue);
    }
}

#[then(expr = "issue {string} should depend on {string} with type {string}")]
fn then_issue_should_depend_on(
    world: &mut KanbusWorld,
    identifier: String,
    target: String,
    dependency_type: String,
) {
    let project_dir = load_project_dir(world);
    let issue = read_issue_file(&project_dir, &identifier);
    assert!(issue
        .dependencies
        .iter()
        .any(|link| link.target == target && link.dependency_type == dependency_type));
}

#[then(expr = "issue {string} should not depend on {string} with type {string}")]
fn then_issue_should_not_depend_on(
    world: &mut KanbusWorld,
    identifier: String,
    target: String,
    dependency_type: String,
) {
    let project_dir = load_project_dir(world);
    let issue = read_issue_file(&project_dir, &identifier);
    assert!(!issue
        .dependencies
        .iter()
        .any(|link| link.target == target && link.dependency_type == dependency_type));
}

#[then(expr = "issue {string} should have 1 dependency")]
fn then_issue_has_single_dependency(world: &mut KanbusWorld, identifier: String) {
    let project_dir = load_project_dir(world);
    let issue = read_issue_file(&project_dir, &identifier);
    assert_eq!(issue.dependencies.len(), 1);
}
